src/api/api.js

```diff
@@ -36,78 +35,84 @@ api_schema.register_api(require('./func_api'));
-        this.rpc = rpc;
-        this.options = _.create(default_options);
-        this.RPC_BUFFERS = RPC_BUFFERS;
-
-        // define the client properties
-        this.auth = undefined;
-        this.account = undefined;
-        this.system = undefined;
-        this.tier = undefined;
-        this.node = undefined;
-        this.host = undefined;
-        this.bucket = undefined;
-        this.events = undefined;
-        this.object = undefined;
-        this.agent = undefined;
-        this.block_store = undefined;
-        this.stats = undefined;
-        this.scrubber = undefined;
-        this.debug = undefined;
-        this.redirector = undefined;
-        this.tiering_policy = undefined;
-        this.pool = undefined;
-        this.cluster_server = undefined;
-        this.cluster_internal = undefined;
-        this.server_inter_process = undefined;
-        this.hosted_agents = undefined;
-        this.frontend_notifications = undefined;
-        this.func = undefined;
-        this.func_node = undefined;
-
-        _.each(rpc.schema, api => {
-            if (!api || !api.id || api.id[0] === '_') return;
-            const name = api.id.replace(/_api$/, '');
-            if (name === 'rpc' || name === 'options') throw new Error('ILLEGAL API ID');
-            this[name] = {};
-            _.each(api.methods, (method_api, method_name) => {
-                this[name][method_name] = (params, options) => {
-                    options = _.create(this.options, options);
-                    return rpc._request(api, method_api, params, options);
-                };
-            });
-        });
-    }
+const client_factory = client_factory_from_schema(api_schema);
 
-    /**
-     * extend the rpc client prototype with convinient methods
-     */
-    create_auth_token(params) {
-        return this.auth.create_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+function client_factory_from_schema(schema) {
+    const client_proto = {
+        RPC_BUFFERS,
 
-    create_access_key_auth(params) {
-        return this.auth.create_access_key_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+        async create_auth_token(params) {
+            const res = this.auth.create_auth(params);
+            this.options.auth_token = res.token;
+            return res;
+        },
+
+        async create_access_key_auth(params) {
+            const res = this.auth.create_access_key_auth(params);
+            this.options.auth_token = res.token;
+            return res;
+        },
+
+        async create_k8s_auth(params) {
+            const res = await this.auth.create_k8s_auth(params);
+            this.options.auth_token = res.token;
+            return res;
+        },
+
+        _invoke_api(api, method_api, params, options) {
```

Comment:
why not async?

Reply:
I can, as a marking.
But please note that adding the async keyword without a need will cause another nextTick schedule instead of returning immediately.

Reply:
but rpc._request always returns a promise so it shouldn't reschedule twice, right?

Reply:
It will reschedule twice. Even If V8 had an optimization to understand that the return value will always be a promise and not do a second reschedules it will not be able to do it in our case because we do not return native promises from RPC, we return bluebird promises (and lose a lot of the new optimization V8 introduce to the compilation of promise-based code).

We want to just pass through the promise so let just drop the async for now

Reply:
Maybe I'm missing something, if I define an `async function` it still doesn't schedule anything, only `await` will do that, and since the function body is still just returning a promise from another function, then isn't `async` only a declarative annotation that makes this function always convert its return value to a promise?

Reply:
On a second note, we can change rpc _request to return native promise too, shouldn't be difficult at all.

Reply:
I will try to clarify. By spec, every function marked as async returns a promise so even:
```
function Async() { return 1; }
```
will return a promise.

Now the V8 engine does not know about alternate implementations of the Promise interface so in order to guarantee a return value of a promise it will wrap whatever you give him with a promise. If the return value is a "thanable" it will chain to it, which is the case with bluebird. The reason that not every "thanable" is treated as a promise is that V8 it cannot guarantee compliance with the spec. 

So even:
```
async function () { return P.resolve() }  
```
will be rescheduled twice, once internally inside bluebord using `nextTick` and once internally inside V8

Reply:
Regarding the rpc_request change.  It's a good direction head to but even with a native promise, there is no guarantee that the optimization will kick in. There are reasons that are far too complex to explain here. I will try to find the V8 blog article and link.

ESLint has a rule that complains about wrong uses, for these exact reasons, of the async keyword but we choose to turn it off.

Reply:
OK thanks. We should get rid of bluebird whenever we can.